doc_internal/rust-devel.md

```diff
@@ -0,0 +1,162 @@
+    "fish"L // equivalent to L!("fish")
+}
+```
+
+### Strings for FFI
+
+`WString` and `&wstr` are the common strings used by Rust components. At the FII boundary there are some additional strings for interop. _All of these are temporary for the duration of the port._
+
+- `CxxWString` is the Rust binding of `std::wstring`. It is the wide-string analog to [`CxxString`](https://cxx.rs/binding/cxxstring.html) and is [added in our fork of cxx](https://github.com/ridiculousfish/cxx/blob/fish/src/cxx_wstring.rs). This is useful for functions which return e.g. `const wcstring &`.
+- `W0String` is renamed [U32CString](https://docs.rs/widestring/latest/widestring/ucstring/struct.U32CString.html). This is basically `WString` except it _is_ nul-terminated. This is useful for getting a nul-terminated `const wchar_t *` to pass to C++ implementations.
+- `wcharz_t` is an annoying C++ struct which merely wraps a `const wchar_t *`, used for passing these pointers from C++ to Rust. We would prefer to use `const wchar_t *` directly but `autocxx` refuses to generate bindings for types such as `std::vector<const wchar_t *>` so we wrap it in this silly struct.
+
+Note C++ `wchar_t`, Rust `char`, and `u32` are effectively interchangeable: you can cast pointers to them back and forth (except we check upon u32->char conversion). However be aware of which types are nul-terminated.
+
+These types should be confined to the FFI modules, in particular `wchar_ffi`. They should not "leak" into other modules. See the `wchar_ffi` module.
+
+### Format strings
+
+Rust's builtin `std::fmt` modules do not accept runtime-provided format strings, so we mostly won't use them, except perhaps for FLOG / other non-translated text.
+
+Instead we'll continue to use printf-style strings, with a Rust printf implementation.
+
+### Vectors
+
+See [`Vec`](https://cxx.rs/binding/vec.html) and [`CxxVector`](https://cxx.rs/binding/cxxvector.html).
+
+In many cases, `autocxx` refuses to allow vectors of certain types. For example, autocxx supports `std::vector` and `std::shared_ptr` but NOT `std::vector<std::shared_ptr<...>>`. To work around this one can create a helper (pointer, length) struct. Example:
+
+```cpp
+struct RustFFIJobList {
+    std::shared_ptr<job_t> *jobs;
+    size_t count;
+};
+```
+
+This is just a POD (plain old data) so autocxx can generate bindings for it. Then it is trivial to convert it to a Rust slice:
+
+```
+pub fn get_jobs(ffi_jobs: &ffi::RustFFIJobList) -> &[SharedPtr<job_t>] {
+    unsafe { slice::from_raw_parts(ffi_jobs.jobs, ffi_jobs.count) }
+}
+```
+
+## Development Tooling
+
+The [autocxx guidance](https://google.github.io/autocxx/workflow.html#how-can-i-see-what-bindings-autocxx-has-generated) is helpful:
+
+1. Install cargo expand (`cargo install cargo-expand`). Then you can use `cargo expand` to see the generated Rust bindings for C++. In particular this is useful for seeing failed expansions for C++ types that autocxx cannot handle.
```

Comment:
`cargo expand` requires nightly rustc as it uses an unstable feature of rustc to perform the macro expansion. Rust-analyzer also has the ["Expand Macro Recursively"](https://rust-analyzer.github.io/manual.html#expand-macro-recursively) command which you can invoke on a single macro invocation. This also works on stable as it uses rust-analyzer's own macro expander rather than the rustc one.

Reply:
It might be better to change these recommended commands into
`cargo +nightly install command` and `cargo +nightly expand`. Rustup should prompt the user to install the nightly toolchain if it is missjng.

I think that `cargo expand` will automatically select the nightly toolchain if it is only installed there.